Gems/PhysX/Code/Source/HeightfieldColliderComponent.cpp

```diff
@@ -62,22 +63,83 @@ namespace PhysX
     {
     }
 
+
+    void HeightfieldColliderComponent::OnAssetReady(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        if (asset == m_bakedHeightfieldAsset)
+        {
+            m_bakedHeightfieldAsset = asset;
+
+            Pipeline::HeightFieldAsset* heightfieldAsset = m_bakedHeightfieldAsset.Get();
+            
+            bool minMaxHeightsMatch = AZ::IsClose(m_shapeConfig->GetMinHeightBounds(), heightfieldAsset->GetMinHeight()) &&
+                AZ::IsClose(m_shapeConfig->GetMaxHeightBounds(), heightfieldAsset->GetMaxHeight());
+
+            if (!minMaxHeightsMatch)
+            {
+                AZ_Warning(
+                    "PhysX",
+                    false,
+                    "MinMax heights mismatch between baked heightfield and heightfield provider. Entity [%s]. "
+                    "Terrain [%0.2f, %0.2f], Asset [%0.2f, %0.2f]",
+                    GetEntity()->GetName().c_str(),
+                    m_shapeConfig->GetMinHeightBounds(),
+                    m_shapeConfig->GetMaxHeightBounds(),
+                    heightfieldAsset->GetMinHeight(),
+                    heightfieldAsset->GetMaxHeight());
+            }
+
+            physx::PxHeightField* pxHeightfield = heightfieldAsset->GetHeightField();
+
+            // Since PxHeightfield will have shared ownership in both HeightfieldAsset and HeightfieldShapeConfiguration,
+            // we need to increment the reference counter here. Both of these places call release() in destructors,
+            // so we need to avoid double deletion this way.
+            pxHeightfield->acquireReference();
+            m_shapeConfig->SetCachedNativeHeightfield(pxHeightfield);
+
+            InitHeightfieldCollider(HeightfieldCollider::DataSource::UseCachedHeightfield);
+        }
+    }
+
+    void HeightfieldColliderComponent::OnAssetReload(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        if (asset == m_bakedHeightfieldAsset)
+        {
+            m_heightfieldCollider.reset();
+            OnAssetReady(asset);
+        }
+    }
+
+    void HeightfieldColliderComponent::OnAssetError(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        InitHeightfieldCollider(HeightfieldCollider::DataSource::GenerateNewHeightfield);
+    }
+
     void HeightfieldColliderComponent::Activate()
     {
-        const AZ::EntityId entityId = GetEntityId();
+        *m_shapeConfig = Utils::CreateBaseHeightfieldShapeConfiguration(GetEntityId());
```

Comment:
What is `m_shapeConfig` and what does `CreateBaseHeightfieldShapeConfiguration` return? This assignment is bit odd dereferencing `*m_shapeConfig`, if it's a smart pointer, where was it created first?

Reply:
`CreateBaseHeightfieldShapeConfiguration` returns `Physics::HeightfieldShapeConfiguration` which we set to an object `m_shapeConfig` pointer refers to. 

This is to initialize HeightfieldShapeConfiguration with the actual data from the terrain system.

m_shapeConfig used to be reflected in the component, so the data was transferred from the EditoHeightfieldComponent, but it was removed from the serializer recently, so was left constructed but uninitialized. 
Now we do initialize it